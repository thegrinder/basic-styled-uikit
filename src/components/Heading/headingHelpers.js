import { getStyle } from '../../helpers/utils';

const getHeadingSizing = ({ theme, sizing }) => theme.uiKit.heading.sizings[sizing];
const headingSizingStyles = getStyle(getHeadingSizing);

export default headingSizingStyles;
