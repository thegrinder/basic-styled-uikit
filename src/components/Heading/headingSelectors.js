import { getHeading } from '../../theme/themeSelectors';

export const getHeadingSizingStyle = (theme, sizing) =>
  getHeading(theme).sizings[sizing] || {};

export const getHeadingColor = (theme, color) =>
  getHeading(theme).colors[color];

export const getHeadingCommonStyle = (theme) => getHeading(theme).common;
