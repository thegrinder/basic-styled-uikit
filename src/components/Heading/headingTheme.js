import colors from '../../theme/colors/colorsTheme';

const headingTheme = {
  sizings: {
    h1: {
      fontSize: '3rem',
    },
    h2: {
      fontSize: '2.5rem',
    },
    h3: {
      fontSize: '2rem',
    },
    h4: {
      fontSize: '1.5rem',
    },
    h5: {
      fontSize: '1.25rem',
    },
    h6: {
      fontSize: '1rem',
    },
  },
  colors: {
    white: colors.white,
    light: colors.gray700,
    default: colors.gray800,
    dark: colors.gray900,
    black: colors.black,

    primary: colors.primary500,
    success: colors.success500,
    warning: colors.warning500,
    danger: colors.danger500,
  },
  common: {
    fontWeight: 700,
    fontFamily: 'Lato, Arial, Helvetica, sans-serif',
  },
};

export default headingTheme;
