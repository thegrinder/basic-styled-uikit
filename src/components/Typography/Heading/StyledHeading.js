import styled, { css } from 'styled-components';
import { string, bool } from 'prop-types';
import { getBaseLineHeight, calcLineHeight } from '../typographySelectors';
import {
  getHeadingSizingStyle,
  getHeadingColor,
  getHeadingColorOnBg,
  getHeadingCommonStyle,
  getHeadingOpacity,
} from './headingSelectors';

const propTypes = {
  color: string.isRequired,
  sizing: string.isRequired,
  marginBottom: bool.isRequired,
  emphasis: string.isRequired,
  ignoreBackground: bool.isRequired,
  bgColor: string,
};

const StyledHeading = styled.h1`
  display: block;
  ${({
    theme,
    color,
    sizing,
    marginBottom,
    emphasis,
    ignoreBackground,
    bgColor,
  }) => {
    const sizingStyle = getHeadingSizingStyle(theme, sizing);
    return css`
      ${getHeadingCommonStyle(theme, sizing)}
      color: ${
        ignoreBackground
          ? getHeadingColor(theme, color)
          : getHeadingColorOnBg(theme, bgColor, color)
      };
      line-height: ${calcLineHeight(sizingStyle?.fontSize)};
      margin-bottom: ${marginBottom ? `${getBaseLineHeight(theme)}rem` : '0'};
      opacity: ${getHeadingOpacity(theme, emphasis)};
      ${sizingStyle}
    `;
  }}
`;

StyledHeading.propTypes = propTypes;

export default StyledHeading;
