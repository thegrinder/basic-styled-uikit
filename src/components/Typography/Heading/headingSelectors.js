import { getTypography } from '../../../theme/themeSelectors';

export const getHeadingSizingStyle = (theme, sizing) => {
  const sizingStyle = getTypography(theme).heading.sizings[sizing];
  if (!sizingStyle) {
    throw new Error(`There is no ${sizing} size in the <Heading /> theme`);
  }
  return sizingStyle;
};

export const getHeadingColor = (theme, color) => {
  let headingColor;

  const baseColors = getTypography(theme).colors;
  if (baseColors && baseColors[color]) {
    headingColor = baseColors[color];
  }

  const headingColors = getTypography(theme).heading.colors;
  if (headingColors && headingColors[color]) {
    headingColor = headingColors[color];
  }

  if (!headingColor) {
    throw new Error(`There is no ${color} color in the <Heading /> theme`);
  }
  return headingColor;
};

export const getHeadingColorOnBg = (theme, bgColor, color) => {
  let headingColor;

  const baseColors = getTypography(theme).colors;
  if (baseColors && baseColors[color]) {
    headingColor = baseColors[color];
  }

  const baseBgColors = getTypography(theme).onBackground;
  if (baseBgColors && baseBgColors[bgColor] && baseBgColors[bgColor][color]) {
    headingColor = baseBgColors[bgColor][color];
  }

  const headingTheme = getTypography(theme).heading;
  if (
    headingTheme &&
    headingTheme.onBackground &&
    headingTheme.onBackground[bgColor] &&
    headingTheme.onBackground[bgColor][color]
  ) {
    headingColor = headingTheme.onBackground[bgColor][color];
  }

  if (!headingColor) {
    throw new Error(
      `There is no ${color} color set for ${bgColor} background color in typography theme`
    );
  }

  return headingColor;
};

export const getHeadingOpacity = (theme, emphasis) => {
  const textOpacity = getTypography(theme).opacities[emphasis];
  if (!textOpacity) {
    throw new Error(`There is no ${emphasis} opacity in the <Heading /> theme`);
  }
  return textOpacity;
};

export const getHeadingCommonStyle = theme =>
  getTypography(theme).heading.common;
