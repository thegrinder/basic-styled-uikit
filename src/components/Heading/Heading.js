import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledHeading from './StyledHeading';

const propTypes = {
  /** determines the size of the heading */
  sizing: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  /** determines the html tag and size, if sizing prop is not specified */
  as: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']).isRequired,
  /** one of: white, light, default, dark, black, primary, success, warning, danger */
  color: string,
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const defaultProps = {
  color: 'default',
  marginBottom: false,
};

const Heading = ({ sizing, color, marginBottom, as, ...rest }) => (
  <StyledHeading
    as={as}
    sizing={sizing || as}
    marginBottom={marginBottom}
    color={color}
    {...rest}
  />
);

Heading.propTypes = propTypes;
Heading.defaultProps = defaultProps;

export default Heading;
