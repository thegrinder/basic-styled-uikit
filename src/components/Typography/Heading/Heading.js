import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import StyledHeading from './StyledHeading';
import { useColorBoxContext } from '../../ColorBox';

const propTypes = {
  /** determines the size of the heading */
  sizing: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  /** one of: high, normal, low */
  emphasis: string,
  /** determines whether colors defined for the background are ignored */
  ignoreBackground: bool,
  /** determines the html tag and size, if sizing prop is not specified */
  as: oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']).isRequired,
  /** one of: neutral, primary, success, danger, warning */
  color: string,
  /** if true adds a bottom margin according to vertical rhythm */
  marginBottom: bool,
};

const Heading = ({
  color = 'neutral',
  emphasis = 'high',
  marginBottom = false,
  ignoreBackground = false,
  sizing,
  as,
  ...rest
}) => (
  <StyledHeading
    as={as}
    sizing={sizing || as}
    marginBottom={marginBottom}
    color={color}
    emphasis={emphasis}
    ignoreBackground={ignoreBackground}
    bgColor={useColorBoxContext()}
    {...rest}
  />
);

Heading.propTypes = propTypes;

export default Heading;
