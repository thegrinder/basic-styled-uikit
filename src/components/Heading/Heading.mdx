---
name: Heading
menu: Components
---

import { Playground, Props } from 'docz';
import Heading from './Heading';
import ColorBox from '../ColorBox/ColorBox';

# Heading

<Props of={Heading} />


### Heading sizes

<Playground>
  <Heading as="h1">Heading H1</Heading>
  <Heading as="h2">Heading H2</Heading>
  <Heading as="h3">Heading H3</Heading>
  <Heading as="h4">Heading H4</Heading>
  <Heading as="h5">Heading H5</Heading>
  <Heading as="h6">Heading H6</Heading>
  <Heading as="h6" sizing="h1">Big H6 heading</Heading>
  <Heading as="h1" sizing="h6">Small H1 heading</Heading>
</Playground>


### Heading colors

<Playground>
  <Heading as="h4" color="primary">primary</Heading>
  <Heading as="h4" color="danger">danger</Heading>
  <Heading as="h4" color="success">success</Heading>
  <Heading as="h4" color="warning">warning</Heading>
  <Heading as="h4" color="neutral" emphasis="high">neutral emphasis high</Heading>
  <Heading as="h4" color="neutral" emphasis="normal">neutral emphasis normal</Heading>
  <Heading as="h4" color="neutral" emphasis="low">neutral emphasis low</Heading>
</Playground>

### Heading colors on different backgrounds

<Playground>
  <ColorBox bgColor="primary">
    <Heading as="h4">neutral on primary background</Heading>
  </ColorBox>
  <ColorBox bgColor="success">
    <Heading as="h4">neutral on success background</Heading>
  </ColorBox>
  <ColorBox bgColor="danger">
    <Heading as="h4">neutral on danger background</Heading>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Heading as="h4">neutral on warning background</Heading>
  </ColorBox>
  <ColorBox bgColor="warning">
    <Heading as="h4" ignoreBackground={true}>
      background ignored - default black heading
    </Heading>
  </ColorBox>
</Playground>

### Heading with a bottom margin

<Playground>
  <Heading as="h3" marginBottom={true}>Heading with margin</Heading>
  <p>Sample paragraph.</p>
</Playground>