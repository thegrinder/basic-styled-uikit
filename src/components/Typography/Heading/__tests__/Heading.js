import React from 'react';
import { render } from '../../../../test-utils';
import Heading from '../Heading';

const children = 'children';

const renderComponent = (props = {}) =>
  render(
    <Heading as="h1" {...props}>
      {children}
    </Heading>
  );

describe('<Heading />', () => {
  it('should render correctly with default props', () => {
    const { queryByText } = renderComponent();
    expect(queryByText(children)).toBeTruthy();
  });

  it('should render correctly with custom props', () => {
    const { queryByText } = renderComponent({
      as: 'h3',
      color: 'primary',
      marginBottom: true,
    });
    expect(queryByText(children)).toBeTruthy();
  });

  ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].forEach(sizing => {
    it(`should render <${sizing}> tag`, () => {
      const { queryByText } = renderComponent({
        as: sizing,
      });
      expect(queryByText(children).tagName).toEqual(sizing.toUpperCase());
    });
  });

  it('should render should large <h6> heading', () => {
    const { queryByText } = renderComponent({
      sizing: 'h1',
      as: 'h6',
    });
    expect(queryByText(children).tagName).toEqual('H6');
  });
});
