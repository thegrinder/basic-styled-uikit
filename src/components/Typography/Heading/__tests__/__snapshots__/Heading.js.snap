// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should render correctly with custom props 1`] = `
<h3
  class="StyledHeading-sc-1nfjshk-0 cAwPBB"
  color="primary"
>
  children
</h3>
`;

exports[`<Heading /> should render correctly with default props 1`] = `
<h1
  class="StyledHeading-sc-1nfjshk-0 BvWng"
  color="neutral"
>
  children
</h1>
`;
